src/recordlinker/linkage/models.py

```diff
@@ -15,35 +29,125 @@ class Person(Base):
+    the algorithm.  By defining them all upfront, we give the user flexibility in
+    adjusting their algorithm configuration without having to reload the data.
+
+    **HERE BE DRAGONS**: IN A PRODUCTION SYSTEM, THESE ENUMS SHOULD NOT BE CHANGED!!!
+    """
+    BIRTHDATE = 1, "Date of Birth"
+    MRN = 2, "Last 4 chars of MRN"
+    SEX = 3, "Gender"
+    ZIP = 4, "Zip Code"
+    FIRST_NAME = 5, "First 4 chars of First Name"
+    LAST_NAME = 6, "First 4 chars of Last Name"
+
+    def __init__(self, id: int, description: str):
+        self.id = id
+        self.description = description
+
+    def to_value(self, data: dict) -> list[str]:
+        """
+        Given a data dictionary of Patient PII data, return a generator of all
+        possible values for this Key.  Many Keys will only have 1 possible value,
+        but some (like first name) could have multiple values.
+        """
+        if self == BlockingKey.BIRTHDATE:
+            return self._extract_birthdate(data)
+        if self == BlockingKey.MRN:
+            return self._extract_mrn_last_four(data)
+        if self == BlockingKey.SEX:
+            return self._extract_sex(data)
+        if self == BlockingKey.ZIP:
+            return self._extract_zipcode(data)
+        if self == BlockingKey.FIRST_NAME:
+            return self._extract_first_name_first_four(data)
+        if self == BlockingKey.LAST_NAME:
+            return self._extract_last_name_first_four(data)
+        return []
+
+    def _extract_birthdate(self, data: dict) -> list[str]:
+        if "birthdate" in data:
+            val = data["birthdate"]
+            if not isinstance(val, (datetime.date, datetime.datetime)):
+                # if not an instance of date or datetime, try to parse it
+                val = dateutil.parser.parse(str(val))
+            return [val.strftime("%Y-%m-%d")]
+        return []
+
+    def _extract_mrn_last_four(self, data: dict) -> list[str]:
+        if "mrn" in data:
+            return [data["mrn"].strip()[-4:]]
+        return []
+
+    def _extract_sex(self, data: dict) -> list[str]:
+        if "sex" in data:
+            val = str(data["sex"]).lower().strip()
+            if val in ["m", "male"]:
+                return ["m"]
+            elif val in ["f", "female"]:
+                return ["f"]
+            return ["u"]
+        return []
+
+    def _extract_zipcode(self, data: dict) -> list[str]:
+        zipcodes = []
+        for address in data.get("address", []):
+            if isinstance(address, dict):
+                if "zip" in address:
+                    zipcodes.append(str(address["zip"]).strip()[0:5])
+        return zipcodes
+
+    def _extract_first_name_first_four(self, data: dict) -> list[str]:
+        names = []
+        for name in data.get("name", []):
+            if isinstance(name, dict):
+                for given in name.get("given", []):
+                    names.append(str(given)[:4])
+        return names
```

Comment:
Ah yes, this is a good point.  I think in this case we'll have the potential for multiple matches to the same record.  We probably need to yield sets and not lists.  Let me add a test case to verify this.

Reply:
I added a test in this commit https://github.com/CDCgov/RecordLinker/pull/30/commits/406f87bff1e49b88b658bf75779c6851a2360899.  I also changed the signature of BlockingKey.to_value to return a set of strings rather than a list to prevent any double matching issues.